source/fab/engine.py

```diff
@@ -0,0 +1,190 @@
+##############################################################################
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+
+import re
+from pathlib import Path
+from typing import \
+    List, \
+    Mapping, \
+    Tuple, \
+    Type
+from fab.artifact import \
+    Artifact, \
+    FileType, \
+    State, \
+    New, \
+    Unknown, \
+    Analysed, \
+    Compiled, \
+    Linked
+from fab.tasks import Task
+from fab.database import \
+    SqliteStateDatabase, \
+    FileInfoDatabase
+
+
+class PathMap(object):
+    def __init__(self,
+                 pattern: str,
+                 filetype: Type[FileType],
+                 state: Type[State]):
+        self._pattern = pattern
+        self._filetype = filetype
+        self._state = state
+
+    @property
+    def pattern(self):
+        return self._pattern
+
+    @property
+    def filetype(self):
+        return self._filetype
+
+    @property
+    def state(self):
+        return self._state
+
+    def match(self, path: Path) -> bool:
+        matched = False
+        if re.match(self.pattern, str(path)):
+            matched = True
+        return matched
+
+
+class Engine(object):
+    def __init__(self,
+                 workspace: Path,
+                 target: str,
+                 pathmaps: List[PathMap],
+                 taskmap: Mapping[
+                     Tuple[Type[FileType], Type[State]],
+                     Task]) -> None:
```

Comment:
The excessivenes of the `taskmap` type information suggests that there may be room for improvement here. Have a think but no need to implement anything in this change.

Reply:
Agreed - I guess it can probably do something be turned into something like a `List[TaskMap]` in a similar way to the `pathmaps` above it